include/annealing/gaussian_annealing.h

```diff
@@ -0,0 +1,215 @@
+// see <http://www.gnu.org/licenses/>.
+
+
+#ifndef GAUSSIAN_ANNEALING_H
+#define GAUSSIAN_ANNEALING_H
+
+#include <complex>
+
+
+//An implementation of Welford's algorithm for mean and variance.
+template <typename FT>
+std::pair<FT,FT> getMeanVariance(std::vector<FT>& vec) {
+    FT mean = 0, M2 = 0, variance = 0, delta;
+
+    int i=0;
+    typename std::vector<FT>::iterator vecit = vec.begin();
+    for( ; vecit!=vec.end(); vecit++, i++){
+        delta = *vecit - mean;
+        mean += delta / (i + 1);
+        M2 += delta * (*vecit - mean);
+        variance = M2 / (i + 1);
+    }
+
+    return std::pair<FT,FT> (mean, variance);
+}
+
+
+// Compute the first variance a_0 for the starting gaussian
+template <class T1, typename FT>
+void get_first_gaussian(T1 &K, FT radius, FT frac, const vars_g var, FT &error, std::vector<FT> &a_vals) {
+
+    int m = K.num_of_hyperplanes(), dim = var.n, i=0;
+    unsigned int iterations = 0;
+    const int maxiter = 10000;
+    const FT tol = 0.0000001;
+    FT sum, lower = 0.0, upper = 1.0, mid;
+    std::vector <FT> dists(m, 0);
+    Eigen::MatrixXd A = K.get_eigen_mat();
+    Eigen::VectorXd b = K.get_eigen_vec();
+
+    typename std::vector<FT>::iterator disit = dists.begin();
+    for ( ; disit!=dists.end(); disit++, i++)
+        *disit = b(i)/A.row(i).norm();
+
+    // Compute an upper bound for a_0
+    while (iterations < maxiter) {
+        iterations += 1;
+        sum = 0.0;
+        for (typename std::vector<FT>::iterator it = dists.begin(); it != dists.end(); ++it) {
+            sum += std::exp(-upper * std::pow(*it, 2.0)) / (2.0 * (*it) * std::sqrt(M_PI * upper));
+        }
+
+        if (sum > frac * error) {
+            upper = upper * 10;
+        } else {
+            break;
+        }
+    }
+
+    if (iterations == maxiter) {
+        std::cout << "Cannot obtain sharp enough starting Gaussian" << std::endl;
+        exit(-1);
+    }
+
+    //get a_0 with binary search
+    while (upper - lower > tol) {
+        mid = (upper + lower) / 2.0;
+        sum = 0.0;
+        for (typename std::vector<FT>::iterator it = dists.begin(); it != dists.end(); ++it) {
+            sum += std::exp(-mid * std::pow(*it, 2)) / (2 * (*it) * std::sqrt(M_PI * mid));
+        }
+
+        if (sum < frac * error) {
+            upper = mid;
+        } else {
+            lower = mid;
+        }
+    }
+
+    a_vals.push_back((upper + lower) / 2.0);
+    error = (1.0 - frac) * error;
+}
+
+
+// Compute a_{i+1} when a_i is given
+template <class T1, typename FT>
+void get_next_gaussian(T1 K,Point &p, FT a, int N, FT ratio, FT C, vars_g var, std::vector<FT> &a_vals){
+
+    FT last_a = a, last_ratio = 0.1;
+    //k is needed for the computation of the next variance a_{i+1} = a_i * (1-1/d)^k
+    FT k = 1.0;
+    const FT tol = 0.00001;
+    bool done=false, print=var.verbose;
+    std::vector<FT> fn(N,0);
+    std::list<Point> randPoints;
+
+    // Set the radius for the ball walk if it is requested
```

Comment:
this is duplicate with code from get_annealing_schedule.

Reply:
done